crates/rmf_site_editor/src/site/robot_properties.rs

```diff
@@ -28,78 +31,323 @@ pub struct UpdateRobotPropertyKinds {
+        app.world_mut().init_resource::<RobotPropertyRegistry>();
+        app.add_event::<UpdateRobotPropertyKinds>()
+            .add_systems(PostUpdate, update_model_instances::<Robot>);
     }
+}
+
+type InsertDefaultValueFn = fn() -> Result<Value, Error>;
+
+pub struct RobotPropertyRegistration {
+    pub widget: Option<Entity>,
+    pub kinds: HashMap<String, RobotPropertyKindRegistration>,
+}
+
+pub struct RobotPropertyKindRegistration {
+    pub default: InsertDefaultValueFn,
+}
+
+/// This resource keeps track of all the properties that can be configured for a robot.
+#[derive(Resource, Deref, DerefMut)]
+pub struct RobotPropertyRegistry(pub HashMap<String, RobotPropertyRegistration>);
+
+impl FromWorld for RobotPropertyRegistry {
+    fn from_world(_world: &mut World) -> Self {
+        Self(HashMap::new())
+    }
+}
+
+/// Implement this plugin to add a new configurable robot property to the site
+pub struct RobotPropertyPlugin<Property, RecallProperty>
+where
+    Property: RobotProperty,
+    RecallProperty: Recall + Component<Mutability = Mutable> + Default,
+    RecallProperty::Source: RobotProperty,
+{
+    _ignore: std::marker::PhantomData<(Property, RecallProperty)>,
+}
+
+impl<Property, RecallProperty> Default for RobotPropertyPlugin<Property, RecallProperty>
+where
+    Property: RobotProperty,
+    RecallProperty: Recall + Component<Mutability = Mutable> + Default,
+    RecallProperty::Source: RobotProperty,
+{
+    fn default() -> Self {
+        Self {
+            _ignore: Default::default(),
+        }
+    }
+}
 
-    for (entity, robot) in model_properties.iter() {
-        if robot.is_changed() {
-            // Update robot property
-            let value = match retrieve_robot_property::<T>(robot.0.clone()) {
-                Ok((property, value)) => {
-                    commands.entity(entity).insert(property);
-                    value
-                }
-                Err(RobotPropertyError::PropertyNotFound(_)) => {
-                    commands.entity(entity).remove::<T>();
-                    serde_json::Value::Object(Map::new())
-                }
-                Err(_) => continue,
-            };
-
-            // Update robot property kinds
-            update_robot_property_kinds.write(UpdateRobotPropertyKinds {
-                entity,
-                label: property_label.clone(),
-                value,
+impl<Property, RecallProperty> Plugin for RobotPropertyPlugin<Property, RecallProperty>
+where
+    Property: RobotProperty,
+    RecallProperty: Recall + Component<Mutability = Mutable> + Default,
+    RecallProperty::Source: RobotProperty,
+{
+    fn build(&self, app: &mut App) {
+        app.world_mut()
+            .resource_mut::<RobotPropertyRegistry>()
+            .0
+            .insert(
+                Property::label(),
+                RobotPropertyRegistration {
+                    widget: None,
+                    kinds: HashMap::new(),
+                },
```

Comment:
I also wonder if it would make sense to derive `Default` for this struct and just insert its default value